url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/ColliderImpl.java

```diff
@@ -0,0 +1,798 @@
+                if (alarm == null)
+                {
+                    m_threadPool.execute( new SelectorAlarm(runnable) );
+                    break;
+                }
+                else if (m_alarm.compareAndSet(alarm, null))
+                {
+                    alarm.cmp = runnable;
+                    m_threadPool.execute( alarm );
+                    break;
+                }
+            }
+        }
+        else
+            tail.nextSelectorThreadRunnable = runnable;
+    }
+
+    public final void executeInSelectorThreadNoWakeup( SelectorThreadRunnable runnable )
+    {
+        assert( runnable.nextSelectorThreadRunnable == null );
+        final SelectorThreadRunnable tail = s_strTailUpdater.getAndSet( this, runnable );
+        if (tail == null)
+            m_strHead = runnable;
+        else
+            tail.nextSelectorThreadRunnable = runnable;
+    }
+
+    public final void executeInSelectorThreadLater( SelectorThreadRunnable runnable )
+    {
+        assert( runnable.nextSelectorThreadRunnable == null );
+        runnable.nextSelectorThreadRunnable = m_strLater;
+        m_strLater = runnable;
+    }
+
+    public final void executeInThreadPool( ThreadPool.Runnable runnable )
+    {
+        m_threadPool.execute( runnable );
+    }
+
+    public void addAcceptor( Acceptor acceptor ) throws IOException
+    {
+        final ServerSocketChannel serverSocketChannel = ServerSocketChannel.open();
+        serverSocketChannel.configureBlocking( false );
+
+        final ServerSocket socket = serverSocketChannel.socket();
+        socket.setReuseAddress( acceptor.reuseAddr );
+        socket.bind( acceptor.getAddr() );
+
+        SessionSharedData sessionSharedData = getSessionSharedData( acceptor );
+
+        AcceptorImpl acceptorImpl = new AcceptorImpl(
+                this,
+                sessionSharedData.getInputQueueDataBlockCache(),
+                acceptor,
+                sessionSharedData.getJoinMessageMaxSize(),
+                sessionSharedData.getJoinPool(),
+                m_selector,
+                serverSocketChannel );
+
+        IOException ex = null;
+
+        m_lock.lock();
+        try
+        {
+            if (m_stop)
+                ex = new IOException( "Collider stopped." );
+            else if (m_emitters.containsKey(acceptor))
+                ex = new IOException( "Acceptor already registered." );
+            else
+                m_emitters.put( acceptor, acceptorImpl );
+        }
+        finally
+        {
+            m_lock.unlock();
+        }
+
+        if (ex == null)
+            acceptorImpl.start();
+        else
+        {
+            try
+            {
+                serverSocketChannel.close();
+            }
+            catch (final IOException ex1)
+            {
+                /* Should never happen */
+                if (s_logger.isLoggable(Level.WARNING))
+                    s_logger.warning( ex1.toString() );
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 612:621, 763:772.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.